rclpy/rclpy/client.py

```diff
@@ -61,15 +30,53 @@ def __init__(
         self.srv_type = srv_type
         self.srv_name = srv_name
         self.qos_profile = qos_profile
-        self.sequence_number = 0
-        self.response = None
+        # Key is a sequence number, value is an instance of a Future
+        self._pending_requests = {}
         self.callback_group = callback_group
         # True when the callback is ready to fire but has not been "taken" by an executor
         self._executor_event = False
 
     def call(self, req):
-        self.response = None
-        self.sequence_number = _rclpy.rclpy_send_request(self.client_handle, req)
+        """
+        Make a service request and wait for the result.
+
+        Do not call this method in a callback or a deadlock may occur.
+
+        :param req: The service request
+        :return: The service response
+        """
+        event = threading.Event()
+
+        def unblock(future):
+            nonlocal event
+            event.set()
+
+        future = self.call_async(req)
+        future.add_done_callback(unblock)
+
+        event.wait()
+        if future.exception() is not None:
+            raise future.exception()
+        return future.result()
+
+    def call_async(self, req):
+        """
+        Make a service request and asyncronously get the result.
+
+        :return: a Future instance that completes when the request does
+        :rtype: :class:`rclpy.task.Future` instance
+        """
+        sequence_number = _rclpy.rclpy_send_request(self.client_handle, req)
+        future = Future()
+        self._pending_requests[sequence_number] = future
```

Comment:
In order to be defensive should this check / ensure that `sequence_number` is not already in the dictionary?

Reply:
Added check in ece3781bbd0b05cd539683a4bb769e2fdbba6135